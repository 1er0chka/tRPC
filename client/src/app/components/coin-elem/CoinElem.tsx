import React, { useContext } from "react";
import styles from "./CoinElem.module.scss";
import Image from "next/image";
import { CoinInfoContext } from "@/app/provider/CoinInfoContext";

const CoinElem = () => {
  const { coin } = useContext(CoinInfoContext);

  return (
    <div className={styles.title}>
      <div className={styles.coinRank}>{coin.rank}.</div>
      <div className={styles.coinImage}>
        <Image
          alt={""}
          width={1}
          height={1}
          layout={"responsive"}
          src={
            "https://assets.coincap.io/assets/icons/" +
            coin.symbol.toLowerCase() +
            "@2x.png"
          }
        ></Image>
      </div>
      <div className={styles.coinName}>{coin.name}</div>
      <div className={styles.coinSymbol}>{coin.symbol}</div>
    </div>
  );
};

export default CoinElem;
