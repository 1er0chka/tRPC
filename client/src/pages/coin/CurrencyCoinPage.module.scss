@import "src/app/styles/fonts";

.content {
  margin: 3vw 10vw 3vw 10vw;

  @media (max-width: 1440px) {
    margin: 3vw 5vw 3vw 5vw;
  }
  @media (max-width: 1280px) {
    margin: 3vw 0 3vw 2vw;
  }
}

.info {
  display: flex;

  @media (max-width: 768px) {
    flex-direction: column-reverse;
  }
}

.infoLeft {
  width: 54vw;

  @media (max-width: 1440px) {
    width: 72vw;
  }
  @media (max-width: 768px) {
    width: 93vw;
  }
}

.back {
  @include text(500);
  text-decoration: none;
  margin-left: 1vw;
}

.infoRight {
  width: 20vw;
  margin-left: 5vw;
  margin-top: 1.7vw;

  @media (max-width: 768px) {
    width: 50vw;
  }
}

.coinPrice {
  @include title(500);
  margin-bottom: 2vw;
}

.error {
  @include title(500);
  margin-left: 1vw;
  margin-bottom: 1vw;
}

.addButton {
  @include text(500);
  margin-top: 2vw;
  background-color: rgba(255, 255, 255, 0.4);
  border: 0.1vw solid rgba(59, 39, 21, 0.3);
  border-radius: 0.1vw;
  padding: 0.2vw 0.8vw 0.2vw 0.8vw;
  width: 100%;

  &:hover {
    background-color: rgba(255, 255, 255, 0.6);
    cursor: pointer;
  }

  &:disabled {
    background-color: rgba(255, 255, 255, 0.2);
    color: rgba(59, 39, 21, 0.7);
    cursor: auto;
  }
}
