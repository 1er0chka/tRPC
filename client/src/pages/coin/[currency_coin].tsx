import "../../app/styles/globals.css";
import React, { useContext, useEffect, useState } from "react";
import { GetStaticPaths, GetStaticProps } from "next";
import CoinPage from "@/app/components/coin-page/CoinPage";
import Header from "@/app/components/header/Header";
import Service from "@/app/service/Service";
import { useRouter } from "next/router";
import { CoinInfoContext } from "@/app/provider/CoinInfoContext";
import Loading from "@/app/components/loading/Loading";
import styles from "@/pages/coin/CurrencyCoinPage.module.scss";
import Link from "next/link";
import { Coin } from "@/app/service/Types";

export type CurrencyCoinProps = {
  slug: string;
};

const CurrencyCoinPage = ({ slug }: CurrencyCoinProps) => {
  const router = useRouter();
  const { coin, setCoin, interval, setTime, setPrice } =
    useContext(CoinInfoContext);
  const [loading, setLoading] = useState<boolean>(true);
  const [isExist, setExist] = useState<boolean>(false);
  const [popularCoins, setPopularCoins] = useState<Coin[]>([]);

  useEffect(() => {
    const getAssets = async () => {
      if (typeof router.query.currency_coin === "string") {
        Service.getAssetsById(router.query.currency_coin)
          .then((data) => {
            setCoin(data);
            setExist(true);
          })
          .catch((err) => {
            console.error(err);
            setExist(false);
            setLoading(false);
          });
      }
    };
    if (slug != undefined) {
      setLoading(true);
      getAssets();
    }
  }, [slug]);

  useEffect(() => {
    const getHistory = async () => {
      Service.getHistoryById(coin.id, interval).then((data) => {
        const x: string[] = [];
        const y: number[] = [];
        switch (interval) {
          case "m1": {
            data.map((time) => {
              x.push(
                new Date(parseInt(time.time)).getUTCHours().toString() +
                  ":" +
                  new Date(parseInt(time.time)).getUTCMinutes().toString(),
              );
              y.push(parseFloat(time.priceUsd));
            });
            break;
          }
          default: {
            data.map((time) => {
              x.push(
                new Date(parseInt(time.time)).getUTCDay().toString() +
                  "." +
                  new Date(parseInt(time.time)).getUTCMonth().toString() +
                  " " +
                  new Date(parseInt(time.time)).getUTCHours().toString() +
                  ":" +
                  new Date(parseInt(time.time)).getUTCMinutes().toString(),
              );
              y.push(parseFloat(time.priceUsd));
            });
            break;
          }
        }
        setTime(x);
        setPrice(y);
        setLoading(false);
      });
    };

    if (coin.id.length > 0) {
      getHistory();
    }
  }, [coin, interval]);

  useEffect(() => {
    setPopularCoins([]);
    const getAssets = async (id: string) => {
      Service.getAssetsById(id)
        .then((coin) => {
          setPopularCoins((prevCoins) => [...prevCoins, coin]);
        })
        .catch((err) => {
          console.error(err);
          setPopularCoins([]);
        });
    };
    getAssets("bitcoin");
    getAssets("ethereum");
    getAssets("tether");
  }, []);

  return (
    <div>
      <Header coins={popularCoins} />
      {loading ? (
        <Loading />
      ) : (
        <div>
          {isExist ? (
            <CoinPage />
          ) : (
            <div className={styles.content}>
              <div className={styles.error}>Page doesn`t exist</div>
              <Link className={styles.back} href={"/"}>
                {" "}
                ‚Üê Back to table
              </Link>
            </div>
          )}
        </div>
      )}
    </div>
  );
};
export default CurrencyCoinPage;

export const getStaticPaths: GetStaticPaths = async () => {
  return {
    paths: [],
    fallback: true,
  };
};

export const getStaticProps: GetStaticProps<CurrencyCoinProps> = async (
  context,
) => {
  const slug = context.params!.currency_coin as string;
  return {
    props: {
      slug,
    },
  };
};
