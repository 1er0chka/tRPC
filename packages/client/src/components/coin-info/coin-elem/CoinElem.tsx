import React, {useContext} from "react";
import styles from "./CoinElem.module.scss";
import {CoinInfoContext} from "../../../provider/CoinInfoContext";

const CoinElem = () => {
    const {coin} = useContext(CoinInfoContext);

    return (
        <div className={styles.title}>
            <div className={styles.coinRank} data-testid="coin-rank">{coin.rank}.</div>
            <img className={styles.coinImage} alt={""}
                 src={"https://assets.coincap.io/assets/icons/" + coin.symbol.toLowerCase() + "@2x.png"}></img>
            <div className={styles.coinName} data-testid="coin-name">{coin.name}</div>
            <div className={styles.coinSymbol} data-testid="coin-symbol">{coin.symbol}</div>
        </div>
    );
};

export default CoinElem;
