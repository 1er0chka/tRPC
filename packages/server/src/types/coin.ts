export interface Coin {
  id: string;
  name: string;
  rank: string;
  symbol: string;
  supply: string;
  maxSupply: string;
  priceUsd: string;
  changePercent24Hr: string;
  marketCapUsd: string;
}

export interface History {
  priceUsd: string;
  time: string;
  circulatingSupply: string;
  date: string;
}