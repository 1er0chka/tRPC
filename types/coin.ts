export interface Coin {
    id: string;
    name: string;
    rank: string;
    symbol: string;
    supply: string;
    maxSupply: string;
    priceUsd: string;
    changePercent24Hr: string;
    marketCapUsd: string;
}

export const defaultCoin: Coin = {
    changePercent24Hr: "",
    id: "",
    marketCapUsd: "",
    maxSupply: "",
    name: "",
    priceUsd: "",
    rank: "",
    supply: "",
    symbol: "",
};

export interface History {
    priceUsd: string;
    time: string;
    circulatingSupply: string;
    date: string;
}

export interface Portfolio {
    id: string;
    number: number;
    oldPrice: number;
}

export interface PortfolioCoin extends Portfolio {
  newPrice: number;
  difference: number;
  name: string;
  symbol: string
}

export type Sort = "rank" | "priceUsd" | "changePercent24Hr" | "marketCapUsd";
