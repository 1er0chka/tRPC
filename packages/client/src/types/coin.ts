import {Coin} from "../../../../types/coin";

export const defaultCoin: Coin = {
    changePercent24Hr: "",
    id: "",
    marketCapUsd: "",
    maxSupply: "",
    name: "",
    priceUsd: "",
    rank: "",
    supply: "",
    symbol: "",
};

export interface Portfolio {
    id: string;
    number: number;
    oldPrice: number;
}

export interface PortfolioCoin extends Portfolio {
  newPrice: number;
  difference: number;
  name: string;
  symbol: string
}

export type Sort = "rank" | "priceUsd" | "changePercent24Hr" | "marketCapUsd";
