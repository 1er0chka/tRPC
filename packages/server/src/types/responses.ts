import {Coin, History} from "../../../../types/coin";


export interface ICoinsResponse {
    data: Coin[];
}

export interface ICoinResponse {
    data: Coin;
}

export interface IHistoryResponse {
    data: History[];
}
