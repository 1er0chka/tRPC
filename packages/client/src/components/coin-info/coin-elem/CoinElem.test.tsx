import {render, screen} from "@testing-library/react";
import CoinElem from "./CoinElem";
import {Coin} from "../../../../../../types/coin";

const coin:Coin = {
    changePercent24Hr: "-0.8101417214350335",
    id: "bitcoin",
    marketCapUsd: "119179791817.6740161068269075",
    maxSupply: "21000000.0000000000000000",
    name: "Bitcoin",
    priceUsd: "6931.5058555666618359",
    rank: "1",
    supply: "17193925.0000000000000000",
    symbol: "BTC"
}

jest.mock('react', () => {
    const actual = jest.requireActual('react')
    return {
        __esModule: true,
        ...actual,
        useContext: () => {
            return {
                coin
            }
        }
    }
})

describe('Coin Elem Tests', () => {
    test('Snapshot test', () => {
        render(
            <CoinElem/>
        )
        const data = screen.queryByTestId('coin-elem')
        expect(data).toMatchSnapshot()
    })
})