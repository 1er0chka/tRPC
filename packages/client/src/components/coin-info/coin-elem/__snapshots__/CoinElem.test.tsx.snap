// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin Elem Tests Snapshot test 1`] = `
<div
  class="title"
  data-testid="coin-elem"
>
  <div
    class="coinRank"
    data-testid="coin-rank"
  >
    1
    .
  </div>
  <img
    alt=""
    class="coinImage"
    src="https://assets.coincap.io/assets/icons/btc@2x.png"
  />
  <div
    class="coinName"
    data-testid="coin-name"
  >
    Bitcoin
  </div>
  <div
    class="coinSymbol"
    data-testid="coin-symbol"
  >
    BTC
  </div>
</div>
`;
